File: python/427. Construct Quad Tree.py
"""
# Definition for a QuadTree node.
class Node(object):
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""
class Solution(object):
    def construct(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: Node
        """

        def dfs(x, y, offset):
            if offset<= 0: return

            for i in range(x, x+offset):
                for j in range(y, y+offset):
                    if grid[i][j] != grid[x][y]:
                        topLeft = dfs(x, y, offset/2)
                        topRight = dfs(x, y+offset/2, offset/2)
                        bottomLeft = dfs(x+offset/2, y, offset/2)
                        bottomRight = dfs(x+offset/2, y+offset/2, offset/2)
                        return Node(False, False, topLeft, topRight, bottomLeft, bottomRight)
            return Node(grid[x][y], True, None, None, None, None)

        return dfs(0, 0, len(grid))

File: go/src/leetcode/427/427. Construct Quad Tree.go
/**
@author ZhengHao Lou
Date    2022/04/29
*/
package main

type Node struct {
	Val         bool
	IsLeaf      bool
	TopLeft     *Node
	TopRight    *Node
	BottomLeft  *Node
	BottomRight *Node
}

/**
https://leetcode-cn.com/problems/construct-quad-tree/
*/
func construct(grid [][]int) *Node {
	var dfs func(x, y, l int) *Node
	dfs = func(x, y, l int) *Node {
		for i := x; i < x+l; i++ {
			for j := y; j < y+l; j++ {
				if grid[i][j] != grid[x][y] {
					return &Node{
						Val:         true,
						IsLeaf:      false,
						TopLeft:     dfs(x, y, l>>1),
						TopRight:    dfs(x, y+l>>1, l>>1),
						BottomLeft:  dfs(x+l>>1, y, l>>1),
						BottomRight: dfs(x+l>>1, y+l>>1, l>>1),
					}
				}
			}
		}
		return &Node{
			Val:    grid[x][y] == 1,
			IsLeaf: true,
		}
	}
	
	return dfs(0, 0, len(grid))
}
